### ISSUE TITLE
Transactional post-donation email reports failure, actually delivers mail

### ISSUE THREAD

**Issue by GinkgoFJG** [cividesk/com.cividesk.email.sparkpost#8]
Upon making a test donation, I get the following error message:

> An error occurred when CiviCRM attempted to send an email (via SMTP). If you received this error after submitting on online contribution or event registration - the transaction was completed, but we were unable to send the email receipt.
>
> The mail library returned the following error message:
> Sparkpost error: HTTP return code 200. Check https://support.sparkpost.com/customer/en/portal/articles/2140916-extended-error-codes for interpretation.
>
> This is probably related to a problem in your Outbound Email Settings (Administer CiviCRM » System Settings » Outbound Email), OR the FROM email address specifically configured for your contribution page or event. Possible causes are:
>
> Your Sendmail path is incorrect.
> Your Sendmail argument is incorrect.
> The FROM Email Address configured for this feature may not be a valid sender based on your email service provider rules.
> Check this page for more information.

However, the mail is delivered.

CiviCRM version: 4.4.21.
CMS: Drupal 7

I haven't dug into this yet. Just logging it so I don't lose track.

**Comment by nganivet** [comment]
Frank, this happens because SparklPost is returning http code 200 (success), but yet has an 'errors' structure in the response.\

Can you please add in file CRM\Sparkpost, line 95:
        sparkpost_log(print_r($errors, TRUE));
then rerun your test donation and tell me what show up in the sparkpost_log file?

The fix will be immediate once I have this info. Thanks.

**Comment by GinkgoFJG** [comment]
@nganivet: The result returned from the API looks like this:

```
stdClass Object
(
    [errors] => Array
       	(
            [0] => stdClass Object
               	(
                    [message] => transmission created, but with validation errors
                    [description] => message submitted for generation successfully
                    [code] => 2000
               	)

       	)

    [results] => stdClass Object
       	(
            [rcpt_to_errors] => Array
               	(
                    [0] => stdClass Object
                       	(
                            [message] => invalid data format/type
                            [description] => invalid recipient address.email: paul@example.org, elaine@example.org, melissa@example.org
                            [code] => 1300
                       	)

               	)

            [total_rejected_recipients] => 1
            [total_accepted_recipients] => 1
            [id] => 102312599756826188
       	)

)
```

I addressed the root issue in #9; instead of passing a comma-separated string of email addresses, we need to format according to SparkPost's expectations. It looks like a lot of changes, but if you go commit by commit you'll see it's actually pretty straightforward.

I'm leaving this issue open because I think the response handling still needs work. It looks like the response handling is going to get involved enough to merit its own method. I haven't tried to address that in my PR, though.

**Comment by nganivet** [comment]
Thanks Frank, PR much appreciated and merged. Also added more information in case of non-standard error message in a commit this morning. Open to suggestions on how to further enhance the error treatment, but seems good enough for an official 1.0 release by the end of the week now.
